futures-util/src/async_await/random.rs

```diff
@@ -39,16 +36,68 @@ fn random() -> u64 {
             hasher.write_usize(COUNTER.fetch_add(1, Ordering::Relaxed));
             seed = hasher.finish();
         }
-        seed
+        seed as usize
     }
 
-    RNG.with(|rng| {
-        let mut x = rng.get();
-        debug_assert_ne!(x.0, 0);
+    #[cfg(not(feature = "std"))]
+    const fn prng_seed() -> usize {
+        // A deterministic seed is required in absense of TLS or any late init
+        42
+    }
+
+    #[cfg(target_pointer_width = "64")]
+    fn xorshift(mut x: usize) -> (usize, usize) {
+        debug_assert_ne!(x, 0);
         x ^= x >> 12;
         x ^= x << 25;
         x ^= x >> 27;
-        rng.set(x);
-        x.0.wrapping_mul(0x2545_f491_4f6c_dd1d)
-    })
+        (
+            x,
+            x.wrapping_mul(0x2545_f491_4f6c_dd1d),
+        )
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    fn xorshift(mut x: usize) -> (usize, usize) {
+        debug_assert_ne!(x, 0);
+        x ^= x << 13;
+        x ^= x >> 17;
+        x ^= x << 5;
+        (x, x)
+    }
+
+    #[cfg(target_pointer_width = "16")]
+    fn xorshift(mut x: usize) -> (usize, usize) {
+        debug_assert_ne!(x, 0);
+        x ^= x << 7;
+        x ^= x >> 9;
+        x ^= x << 8;
+        (x, x)
+    }
+
+    #[cfg(feature = "std")]
+    fn rng() -> usize {
+        thread_local! {
+            static RNG: Cell<usize> = Cell::new(prng_seed());
+        }
+
+        RNG.with(|rng| {
+            let (x, res) = xorshift(rng.get());
+            rng.set(x);
+            res
+        })
+    }
+
+    #[cfg(not(feature = "std"))]
+    fn rng() -> usize {
+        static RNG: AtomicUsize = AtomicUsize::new(prng_seed());
+
+        // Preemption here can cause multiple threads to observe the same state
```

Comment:
I think this is actually fine-- multiple threads can't be polling the same future at once, so all this means is that the ordering of two unrelated `select!`s will be the same.

Reply:
That's a good point, though I think I framed the issue poorly in the comment. While I don't think it's a big deal because you will always eventually make progress, and the problem seems somewhat convoluted and only really scales if you have tons of threads, what I believe can happen is:

```rust
rng state = 1
thread1: load state = 1
thread2: load state = 1
thread2: shuffle + store state = 2
thread2: load state = 2
thread2: shuffle + store state = 3
thread1: shuffle + store state = 2
thread2: load state = 2 // oh no state was clobbered and the sequence jumped backward from thread2's PoV
thread2: shuffle + store state = 3
```